libsql-server/src/namespace/mod.rs

```diff
@@ -533,80 +575,124 @@ impl<M: MakeNamespace> NamespaceStore<M> {
 
     pub async fn with<Fun, R>(&self, namespace: NamespaceName, f: Fun) -> crate::Result<R>
     where
-        Fun: FnOnce(&Namespace<M::Database>) -> R,
+        Fun: FnOnce(&Namespace<M::Database>) -> R + 'static,
     {
-        if self.inner.has_shutdown.load(Ordering::Relaxed) {
-            return Err(Error::NamespaceStoreShutdown);
-        }
-        let before_load = Instant::now();
-        let lock = self.inner.store.upgradable_read().await;
-        if let Some(ns) = lock.get(&namespace) {
-            Ok(f(ns))
-        } else {
-            let mut lock = RwLockUpgradableReadGuard::upgrade(lock).await;
-            let ns = self
-                .inner
-                .make_namespace
-                .create(
-                    namespace.clone(),
-                    RestoreOption::Latest,
-                    NamespaceBottomlessDbId::NotProvided,
-                    self.inner.allow_lazy_creation,
-                    self.make_reset_cb(),
-                )
-                .await?;
-            let ret = f(&ns);
-            tracing::info!("loaded namespace: `{namespace}`");
-            lock.insert(namespace, ns);
+        let init = {
+            let namespace = namespace.clone();
+            async move {
+                if namespace != NamespaceName::default()
+                    && !self.inner.make_namespace.exists(&namespace)
```

Comment:
Hmmm we may need to check from bottomless if we need a restore, in case the namespace exists but is not on disk

Reply:
The code in master does care about bottomless, the `allow_lazy` creation is passed down to `new_primary`, which makes the decision to create or not. This happens here: https://github.com/tursodatabase/libsql/blob/main/libsql-server/src/namespace/mod.rs#L930.

As you see, we try to go as far as possible with the creation. If the db doesn't exist AND we didn't recover for bottomless AND we're not allowed to lazy create, then it's an error, we abort and clean after ourselves

Reply:
Alright, that makes the last couple of commits counterproductive, I tried to push the check as early as possible to simplify the logic. On the other hand, let my try a different direction: what is the bad effect of us creating a namespace and not returning an error, if it in fact resides in bottomless? From there I see 3 directions to go:
1. Roll back to the previous logic with trying to create and then bailing out, and just make sure the new caching code still works as expected
2. If it's actually harmless, leave as is
3. If harmless, maybe just expand the `exists()` implementation to also check if there's a bottomless laying around somewhere, if enabled? Should be a single http call to s3, and creating a namespace is considered rare